cluster-autoscaler/cloudprovider/azure/azure_cloud_provider.go

```diff
@@ -122,8 +125,15 @@ func (azure *AzureCloudProvider) NodeGroupForNode(node *apiv1.Node) (cloudprovid
 }
 
 // HasInstance returns whether a given node has a corresponding instance in this cloud provider
```

Comment:
Gave this some thought; here is what I _think_ `HasInstance` behavior should be, captured as comment. Let me know if you disagree (avoiding returning `false, nil` = "no instance" is probably the most controversial?) Need to agree on this first, as it would/should naturally be reflected in implementation and tests.

```suggestion
// HasInstance returns whether a given node has a corresponding instance in this cloud provider.
//
// Used to prevent undercount of existing VMs (taint-based overcount of deleted VMs),
// and so should not return false, nil (no instance) if uncertain; return error instead.
// (Think "has instance for sure, else error".) Returning an error causes fallback to taint-based
// determination; use ErrNotImplemented for silent fallback, any other error will be logged.
//
// Expected behavior (should work for VMSS, VMSS Flex and VMs):
// -  exists            : return true, nil
// - !exists            : return *,    NotImplemented (could use custom error for autoscaled nodes)
// - unimplemented case : return *,    NotImplemented
// - any other error    : return *,    error
```